drivers/sensor/lsm6dsr/lsm6dsr.h

```diff
@@ -0,0 +1,112 @@
+/* ST Microelectronics LSM6DSO 6-axis IMU sensor driver
+ *
+ * Copyright (c) 2019 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Datasheet:
+ * https://www.st.com/resource/en/datasheet/lsm6dsr.pdf
+ */
+
+#ifndef ZEPHYR_DRIVERS_SENSOR_LSM6DSR_LSM6DSR_H_
+#define ZEPHYR_DRIVERS_SENSOR_LSM6DSR_LSM6DSR_H_
+
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/types.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <stmemsc.h>
+#include "lsm6dsr_reg.h"
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+#include <zephyr/drivers/spi.h>
+#endif /* DT_ANY_INST_ON_BUS_STATUS_OKAY(spi) */
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
+#include <zephyr/drivers/i2c.h>
+#endif /* DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c) */
+
+#define LSM6DSR_EN_BIT					0x01
+#define LSM6DSR_DIS_BIT					0x00
+
+/* Accel sensor sensitivity grain is 61 ug/LSB */
+#define SENSI_GRAIN_XL				(61LL)
+
+/* Gyro sensor sensitivity grain is 4375 udps/LSB */
+#define SENSI_GRAIN_G				(4375LL)
+#define SENSOR_PI_DOUBLE			(SENSOR_PI / 1000000.0)
+#define SENSOR_DEG2RAD_DOUBLE		(SENSOR_PI_DOUBLE / 180)
+#define SENSOR_G_DOUBLE				(SENSOR_G / 1000000.0)
+
+struct lsm6dsr_config {
+	stmdev_ctx_t ctx;
+	union {
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(i2c)
```

Comment:
The `#if ... #endif` can be removed here (surely at least one of them is used). I would suggest also using a union of pointers instead of copying the data into the `config`.

Reply:
It is likely the case that at exactly one of them used.  By having both always defined, the union will be size of the larger of the two bus types even when the smaller is the one being used.

Reply:
I'm not sure how I should do this. This is how the config is being populated:
```
                /* if SPI */
		.stmemsc_cfg = {					\
			.spi = SPI_DT_SPEC_INST_GET(inst,		\
					   LSM6DSR_SPI_OP,		\
					   0),				\
		},							\

                /* if I2C */
		.stmemsc_cfg = {					\
			.i2c = I2C_DT_SPEC_INST_GET(inst),		\
		},							\

```

Reply:
That would be one use for `IF_ENABLED()`. It might also be good to factor that out into a separate macro.
```cpp
#define DEFINE_STMEMSC_CFG(inst) ... and then you can call the separate macro from `DEFINE_LSM6DSR()`
```